compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -656,3 +658,44 @@ fn promoted_mir<'tcx>(
 
     tcx.arena.alloc(promoted)
 }
+
+/// Returns the deduced parameter attributes for a function.
+///
+/// Deduced parameter attributes are those that can only be soundly determined by examining the
+/// body of the function instead of just the signature. These can be useful for optimization
+/// purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+/// dependent crates can use them.
+fn deduced_param_attrs<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx [DeducedParamAttrs] {
+    // For now, don't deduce any parameter attributes on intrinsics.
+    //
+    // FIXME(pcwalton): Deduce some?
+    if !tcx.is_closure(def_id) && tcx.is_intrinsic(def_id) {
+        return &[];
+    }
+
+    // Don't deduce any parameter attributes on default methods.
+    //
+    // FIXME(pcwalton): Can we do better here?
+    if tcx.trait_of_item(def_id).is_some() {
+        return &[];
+    }
+
+    // Don't deduce any parameter attributes on foreign items, since they have no MIR.
+    if tcx.is_foreign_item(def_id) {
+        return &[];
+    }
+
+    // Deduced attributes for other crates should be read from the metadata instead of via this
+    // function.
+    debug_assert!(def_id.is_local());
+
+    // Grab the optimized MIR. Set the read-only attribute for every argument declared immutable
+    // that contains no `UnsafeCell`s.
+    let body: &Body<'tcx> = tcx.optimized_mir(def_id);
+    tcx.arena.alloc_from_iter(body.local_decls.iter().skip(1).take(body.arg_count).map(
+        |local_decl| DeducedParamAttrs {
+            read_only: local_decl.mutability == Mutability::Not
```

Comment:
The declared mutability of local decls has no meaning in MIR - in other words, it is not unsound for optimizations to introduce writes to locals that are declared immutable (and as a matter of fact these probably already exist given the likely semantics of move operands). You need to analyze the actual MIR body to find writes instead

Reply:
That's really unfortunate. Is there any will to change that?

Reply:
It is still an open question whether a move is a write in MIR. While that question is open I do not think it makes sense to change this, since if the answer is "yes" (which I expect) then the mutability declaration will be wrong anyway.

If the answer turns out to be "no," then it would in principle be possible to change this. We would need to do two things: 1) enforce this in Miri, and 2) audit existing optimizations for incorrectness under this model. I'm not sure we would actually want this though, but while the move question is open that point is moot anyway.